src/hugr/validate.rs

```diff
@@ -558,6 +521,77 @@ impl<'a, 'b> ValidationContext<'a, 'b> {
         }
         .into())
     }
+
+    fn validate_subtree(
+        &mut self,
+        node: Node,
+        var_decls: &[TypeParam],
+    ) -> Result<(), ValidationError> {
+        let op_type = self.hugr.get_optype(node);
+        // The op_type must be defined only in terms of type variables defined outside the node
+        // TODO consider turning this match into a trait method?
+        match op_type {
+            OpType::LeafOp(crate::ops::LeafOp::CustomOp(b)) => {
+                // Try to resolve serialized names to actual OpDefs in Extensions.
+                let e: Option<ExtensionOp>;
+                let ext_or_opaq = match &**b {
+                    ExternalOp::Opaque(op) => {
+                        // If resolve_extension_ops has been called first, this would always return Ok(None)
+                        e = resolve_opaque_op(node, op, self.extension_registry)?;
+                        e.as_ref().map(Either::Left).unwrap_or(Either::Right(op))
+                    }
+                    ExternalOp::Extension(ext) => Either::Left(ext),
+                };
+                // If successful, check TypeArgs are valid for the declared TypeParams
+                match ext_or_opaq {
+                    Either::Left(exten) => exten
+                        .def()
+                        .validate_args(exten.args(), self.extension_registry, var_decls)
+                        .map_err(|cause| ValidationError::SignatureError { node, cause })?,
+                    Either::Right(opaq) => {
+                        // Best effort. Just check TypeArgs are valid (in themselves, not necessarily wrt the TypeParams)
+                        // and assuming none are binary params (all may contain type vars)
+                        for arg in opaq.args() {
+                            arg.validate(self.extension_registry, var_decls)
+                                .map_err(|cause| ValidationError::SignatureError { node, cause })?;
+                        }
+                    }
+                }
+            }
+            OpType::LeafOp(crate::ops::LeafOp::TypeApply { ta }) => {
+                ta.validate(self.extension_registry)
+                    .map_err(|cause| ValidationError::SignatureError { node, cause })?;
+            }
+            _ => (),
+        }
+
+        // Check port connections.
+        for dir in Direction::BOTH {
+            for (i, port_index) in self.hugr.graph.ports(node.pg_index(), dir).enumerate() {
+                let port = Port::new(dir, i);
+                self.validate_port(node, port, port_index, op_type, var_decls)?;
+            }
+        }
+
+        // Compute cumulative TypeParams for nested FuncDefns, i.e. allow inner FuncDefns
```

Comment:
This is a policy that we could change, but it'd not simplify things much - we'd lose this appending logic, but we'd still need to do the hierarchical traversal, because nodes/ports/edges could still refer to type variables bound by the closest-ancestral FuncDefn node, which could be many levels above in a deep nest of Dfg/Cfg nodes.

Reply:
It'd not simplify things much *unless* we are about to simplify the type hierarchy (and remove polymorphic lambdas), whereupon we might then also be able to remove InsideBinders, Renumber, Substitution, etc....hmmm

Reply:
Simplified.